### PR TITLE
Add binary hardening compile and link flags

### PR DESCRIPTION
**Pull request by thejohnfreeman**
Add full RELRO and stack canary when compiling with GCC or Clang.

### CODE REVIEW

**Comment by nbougalis** [comment]
Awesome! Thanks for adding this, @thejohnfreeman! I wanted to add this for a really long time.

**Review by nbougalis — APPROVED** [review]

**Review by seelabs — APPROVED** [review]

File: Builds/CMake/RippledCompiler.cmake [hunk 1]
```diff
@@ -13,7 +13,6 @@ link_libraries (Ripple::common)
 set_target_properties (common
   PROPERTIES INTERFACE_POSITION_INDEPENDENT_CODE ON)
 set(CMAKE_CXX_EXTENSIONS OFF)
-target_compile_features (common INTERFACE cxx_std_17)
```
Comment (intelliot) [inline Builds/CMake/RippledCompiler.cmake:16]: Why is this line removed?

**Comment by intelliot** [comment]
Suggested commit message:
```
build: add binary hardening compile and link flags (#4603)

Enhance security during the build process:

* The '-fstack-protector' flag enables stack protection for preventing
  buffer overflow vulnerabilities. If an attempt is made to overflow the
  buffer, the program will terminate, thus protecting the integrity of
  the stack.
* The '-Wl,-z,relro,-z,now' linker flag enables Read-only Relocations
  (RELRO), a feature that helps harden the binary against certain types
  of exploits, particularly those that involve overwriting the Global
  Offset Table (GOT).
  * This flag is only set for Linux builds, due to compatibility issues
    with apple-clang.
  * The `relro` option makes certain sections of memory read-only after
    initialization to prevent them from being overwritten, while `now`
    ensures that all dynamic symbols are resolved immediately on program
    start, reducing the window of opportunity for attacks.
```

File: Builds/CMake/RippledCompiler.cmake (hunk 1 above)
Comment (thejohnfreeman) [reply Builds/CMake/RippledCompiler.cmake:16]: Our code should be compiled with at least C++20, and the choice of standard is made by the toolchain file, not our CMake project files.

**Comment by intelliot** [comment]
@thejohnfreeman can you bring this up-to-date with develop?

I don't have permission to apply a merge commit to your branch.

**Pull request merged**